### PR TITLE
Feature Pyramid Networks

### PR DESCRIPTION
**Pull request by Hakuyume**

### CODE REVIEW

File: chainercv/links/model/fpn/head.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            gt_loc = bboxes[i][gt_index].copy()
+        else:
+            gt_loc = xp.empty_like(rois[mask])
+        # tlbr -> yxhw
+        gt_loc[:, 2:] -= gt_loc[:, :2]
+        gt_loc[:, :2] += gt_loc[:, 2:] / 2
+        # offset
+        gt_loc[:, :2] = (gt_loc[:, :2] - rois_yx[mask]) / \
+            rois_hw[mask] / std[0]
+        gt_loc[:, 2:] = xp.log(gt_loc[:, 2:] / rois_hw[mask]) / std[1]
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/head.py]: I find it convenient to see this operation defined in a function.
Could you define one or use the one included already in ChainerCV?
I think your style seems to be simpler than the one below.

https://github.com/chainer/chainercv/blob/master/chainercv/links/model/faster_rcnn/utils/bbox2loc.py

**Comment by Hakuyume** [comment]
train/eval/demo scripts are working now.
I will add docs and tests.

**Comment by Hakuyume** [comment]
Although we discussed `loc2bbox/bbox2loc` in https://github.com/chainer/chainercv/pull/685#discussion_r231116336, I want to add that in another PR.

**Comment by yuyu2172** [comment]
OK

File: chainercv/links/model/fpn/head.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+
+    Args:
+        n_class (int): The number of classes including background.
+        scales (tuple of floats): The scales of feature maps.
+
+    """
+    _canonical_level = 2
+    _canonical_scale = 224
+    _roi_size = 7
+    _roi_sample_ratio = 2
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/head.py]: Is `_roi_sample_ratio` used?

File: chainercv/links/model/fpn/head.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+    Args:
+        rois (iterable of arrays): An iterable of arrays of
+            shape :math:`(R_l, 4)`, where :math:`R_l` is the number
+            of RoIs in the :math:`l`-th feature map.
+        roi_indices (iterable of arrays): An iterable of arrays of
+            shape :math:`(R_l,)`.
+        std (tuple of floats): Two coefficients used for encoding
+            bounding boxes.
+        bboxes (list of arrays): A lisf of arrays whose shape is
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/head.py]: lisf -> list

File: chainercv/links/model/fpn/head.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            :math:`(R_n, 4)`, where :math:`R_n` is the number of
+            ground truth bounding boxes.
+        labels (list of arrays): A lisf of arrays whose shape is
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/head.py]: ditto

File: chainercv/links/model/fpn/rpn.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+             :math:`(N, K_l, 4)`, where :math:`K_l` is the number of
+             the anchor boxes of the :math:`l`-th level.
+         confs (iterable of arrays): An iterable of arrays whose shape is
+             :math:`(N, K_l)`.
+         anchors (list of arrays): A list of arrays returned by
+             :meth:`anchors`.
+         sizes (list of tuples of two ints): A list of
+             :math:`(H_n, W_n)`, where :math:`H_n` and :math:`W_n`
+             are height and width of the :math:`n`-th image.
+         bboxes (list of arrays): A lisf of arrays whose shape is
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/rpn.py]: ditto

File: chainercv/links/model/fpn/head.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+            :math:`(R_n,)`.
+
+     Returns:
+         tuple of four arrays:
```
Comment (yuyu2172) [inline chainercv/links/model/fpn/head.py]: tuple of four lists?

File: examples/fpn/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+## Performance
+MS COCO 2017 Val
+
+| Model | Original | ChainerCV |
+|:-:|:-:|:-:|:-:|
+| Faster-RCNN FPN ResNet50 | 36.7 % [2] | 37.1 % |
+| Faster-RCNN FPN ResNet50 | 39.4 % [2] | 39.5 % |
+
+Scores are mean of mean Average Precision (mmAP).
```
Comment (yuyu2172) [inline examples/fpn/README.md]: the mean of mean Average Precision

File: examples/fpn/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+```
+
+## Evaluation
+The evaluation can be conducted using [`chainercv/examples/detection/eval_coco.py`](https://github.com/chainer/chainercv/blob/master/examples/detection).
+
+## Train
+You can train the model with the following code.
+Note that this code requires `chainermn` module.
+```
+$ mpi4exec -n <#gpu> python train_multi.py [--model faster_rcnn_fpn_resnet50|faster_rcnn_fpn_101] [--batchsize <batchsize>]
```
Comment (yuyu2172) [inline examples/fpn/README.md]: mpiexec?

File: examples_tests/detection_tests/test_eval_coco.sh [hunk 9]
```diff
@@ -0,0 +1,5 @@
+cd examples/detection
+sed -e 's/return_crowded=True)/return_crowded=True).slice[:20]/' -i eval_coco.py
+
+$PYTHON eval_coco.py --model faster_rcnn_fpn_resnet50 --gpu 0 --batchsize 2
+$PYTHON eval_coco.py --model faster_rcnn_fpn_resnet10 --gpu 0 --batchsize 2
```
Comment (yuyu2172) [inline examples_tests/detection_tests/test_eval_coco.sh]: typo?

File: examples/fpn/README.md (hunk 8 above)
Comment (Hakuyume) [reply examples/fpn/README.md]: README(s) of other models also use `mpi4exec`. So, I guess this is ok.

Comment (Hakuyume) [reply examples/fpn/README.md]: I changed my mind since the official document of ChainerMN uses `mpiexec`. Should I update all README(s) in this PR?

Comment (yuyu2172) [reply examples/fpn/README.md]: It is ok to change all READMEs in this PR.

**Pull request merged**